swarm-test/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+use async_trait::async_trait;
+use futures::future::Either;
+use futures::StreamExt;
+use libp2p::core::transport::MemoryTransport;
+use libp2p::core::upgrade::Version;
+use libp2p::identity::Keypair;
+use libp2p::noise::NoiseAuthenticated;
+use libp2p::swarm::{
+    AddressScore, ConnectionHandler, IntoConnectionHandler, NetworkBehaviour, SwarmEvent,
+};
+use libp2p::yamux::YamuxConfig;
+use libp2p::{Multiaddr, PeerId, Swarm, Transport};
+use std::fmt::Debug;
+use std::time::Duration;
+
+type THandler<TBehaviour> = <TBehaviour as NetworkBehaviour>::ConnectionHandler;
+type THandlerErr<TBehaviour> =
+    <<THandler<TBehaviour> as IntoConnectionHandler>::Handler as ConnectionHandler>::Error;
+
+/// An extension trait for [`Swarm`] that makes it easier to set up a network of [`Swarm`]s for tests.
+#[async_trait]
+pub trait SwarmExt {
+    type NB: NetworkBehaviour;
+
+    /// Create a new [`Swarm`] with an ephemeral identity.
+    ///
+    /// The swarm will use a [`MemoryTransport`] together with a noise authentication layer and
+    /// yamux as the multiplexer. However, these details should not be relied upon by the test
+    /// and may change at any time.
+    fn new_ephemeral(behaviour_fn: impl FnOnce(PeerId, Keypair) -> Self::NB) -> Self
```

Comment:
```suggestion
    fn new_ephemeral(behaviour_fn: impl FnOnce(Keypair) -> Self::NB) -> Self
```
Since the `PeerId` can be derived from the `Keypair`, what do you think about removing it as parameter?

Reply:
We could actually, yeah. It is a bit verbose though and many times, behaviours only need to know the `PeerId` and not the `Keypair`. Perhaps I can make some clever abstractions here so the user can choose ...

Reply:
Addressed in [`c6b670d` (#2888)](https://github.com/libp2p/rust-libp2p/pull/2888/commits/c6b670d19b4a4c27cab660bc12688aa6578dd69c).